### PR TITLE
 [LFRic] Support for r_tran_operator_type (#1998)

### PR DESCRIPTION
**Pull request by TeranIvy**
Enhanced `lfric_constants.py` and brought the copy of the LFRic operator infrastructure up-to-date with LFRic.

Updated documentation accordingly and added tests.

This support is required for the next PSyclone release.

### CODE REVIEW

File: doc/user_guide/dynamo0p3.rst [hunk 1]
```diff
@@ -431,13 +431,18 @@ associated kernel metadata description and their precision:
 +--------------------------+------------------------+-----------+
 | R_SOLVER_OPERATOR_TYPE   | GH_OPERATOR            | R_SOLVER  |
 +--------------------------+------------------------+-----------+
+| R_TRAN_OPERATOR_TYPE     | GH_OPERATOR            | R_TRAN    |
++--------------------------+------------------------+-----------+
 | COLUMNWISE_OPERATOR_TYPE | GH_COLUMNWISE_OPERATOR | R_SOLVER  |
 +--------------------------+------------------------+-----------+
 
+Note, LMA and CMA operators can only have ``GH_REAL`` (:ref:`argument
```
Comment (arporter) [inline doc/user_guide/dynamo0p3.rst:439]: I must be missing something here: the table does contain LMA (`OPERATOR_TYPE`) and CMA (`COLUMNWISE_OPERATOR_TYPE`) entries. What am I missing (apart from coffee)?

File: doc/user_guide/dynamo0p3.rst [hunk 2]
```diff
@@ -615,21 +621,23 @@ PSyclone must be able to determine the datatype of an LMA operator.
...
+of a column-wise (CMA) operator. The reason for this is that only one
 datatype is supported, a ``columnwise_operator_type`` which contains
-real data with precision ``r_solver``. PSyclone can therefore simply add
-this datatype in the PSy-layer. However, if the datatype information
-is found in the algorithm layer then and it is not of the expected
-type then PSyclone will abort with a message that indicates the
-problem.
+``real``-valued data with precision ``r_solver``. PSyclone can therefore
+simply add this datatype in the PSy-layer. However, if the datatype
+information is found in the algorithm layer then and it is not of the
```
Comment (arporter) [inline doc/user_guide/dynamo0p3.rst:638]: Not yours but s/then and/and/

File: src/psyclone/tests/dynamo0p3_test.py [hunk 3]
```diff
@@ -1897,6 +1897,44 @@ class for an r_solver_operator.
             "algorithm code." in str(info.value))
 
 
+def test_dynkernelargument_idtp_r_tran_operator(tmpdir):
+    '''
+    Test the _init_data_type_properties method in the DynKernelArgument
+    class for an r_tran_operator.
+
+    '''
+    # Use one of the examples to create an instance of
```
Comment (arporter) [inline src/psyclone/tests/dynamo0p3_test.py:1906]: s/examples/test algorithms/ perhaps?

**Review by arporter — CHANGES_REQUESTED** [review]
Nicely done Iva.
All tests and examples are fine with compilation.
Updated UG builds fine.
Reference guide still builds fine.
Issue number is not mentioned anywhere.
Just a couple of minor things to fix (and please bring up-to-date with master).

File: doc/user_guide/dynamo0p3.rst (hunk 2 above)
Comment (TeranIvy) [reply doc/user_guide/dynamo0p3.rst:638]: Done.

File: src/psyclone/tests/dynamo0p3_test.py (hunk 3 above)
Comment (TeranIvy) [reply src/psyclone/tests/dynamo0p3_test.py:1906]: Yes, more consistent, thanks. Changed it here and in similar tests.

File: doc/user_guide/dynamo0p3.rst (hunk 1 above)
Comment (TeranIvy) [reply doc/user_guide/dynamo0p3.rst:439]: All other entries under Kernel Metadata column have argument type and data type (e.g. `GH_FIELD, GH_REAL`). This is omitted for operators. I agree, the note is confusing so I removed it and updated the operator entries with `GH_REAL` and re-aligned everything instead. I checked the build, the table is still within the text margins.

**Comment by TeranIvy** [comment]
Speedy @arporter strikes again - thank you! 😃 Addressed the review comments and updated the branch as requested.

File: doc/user_guide/dynamo0p3.rst (hunk 1 above)
Comment (arporter) [reply doc/user_guide/dynamo0p3.rst]: Oh I see. Thanks for updating.

File: src/psyclone/tests/dynamo0p3_test.py (hunk 3 above)
Comment (arporter) [reply src/psyclone/tests/dynamo0p3_test.py]: Above and beyond, thanks :-)

**Review by arporter — APPROVED** [review]
All requested changes have been made.
Tests and examples are still all OK.
Ref and User Guides are both OK.
Will proceed to merge.

**Pull request merged**